// @zennolab_com/capmonstercloud-client/dist/CapMonsterModules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CapMonsterModules = void 0;
/**
 * capmonster.cloud recognition modules
 * {@link https://zennolab.atlassian.net/wiki/spaces/APIS/pages/187269121/How+to+pass+module+name+to+CapMonster+Cloud+using+ApiKey+field+only}
 */
var CapMonsterModules;
(function (CapMonsterModules) {
    CapMonsterModules["Amazon"] = "amazon";
    CapMonsterModules["BotDetect"] = "botdetect";
    CapMonsterModules["Facebook"] = "facebook";
    CapMonsterModules["Gmx"] = "gmx";
    CapMonsterModules["Google"] = "google";
    CapMonsterModules["Hotmail"] = "hotmail";
    CapMonsterModules["MailRu"] = "mailru";
    CapMonsterModules["Ok"] = "ok";
    CapMonsterModules["OkNew"] = "oknew";
    CapMonsterModules["RamblerRus"] = "ramblerrus";
    CapMonsterModules["SolveMedia"] = "solvemedia";
    CapMonsterModules["Steam"] = "steam";
    CapMonsterModules["Vk"] = "vk";
    CapMonsterModules["Yandex"] = "yandex";
    /**
     * Yandex (two words)
     */
    CapMonsterModules["YandexNew"] = "yandexnew";
    CapMonsterModules["YandexWave"] = "yandexwave";
    /**
     * All other text captcha types
     */
    CapMonsterModules["Universal"] = "universal";
})(CapMonsterModules = exports.CapMonsterModules || (exports.CapMonsterModules = {}));
