// @zennolab_com/capmonstercloud-client/dist/TaskType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskType = void 0;
var TaskType;
(function (TaskType) {
    TaskType["FunCaptchaTaskProxyless"] = "FunCaptchaTaskProxyless";
    TaskType["FunCaptchaTask"] = "FunCaptchaTask";
    TaskType["GeeTestTaskProxyless"] = "GeeTestTaskProxyless";
    TaskType["GeeTestTask"] = "GeeTestTask";
    TaskType["HCaptchaTaskProxyless"] = "HCaptchaTaskProxyless";
    TaskType["HCaptchaTask"] = "HCaptchaTask";
    TaskType["ImageToTextTask"] = "ImageToTextTask";
    TaskType["RecaptchaV2EnterpriseTaskProxyless"] = "RecaptchaV2EnterpriseTaskProxyless";
    TaskType["RecaptchaV2EnterpriseTask"] = "RecaptchaV2EnterpriseTask";
    TaskType["NoCaptchaTaskProxyless"] = "NoCaptchaTaskProxyless";
    TaskType["NoCaptchaTask"] = "NoCaptchaTask";
    TaskType["RecaptchaV3TaskProxyless"] = "RecaptchaV3TaskProxyless";
    TaskType["TurnstileTask"] = "TurnstileTask";
    TaskType["TurnstileTaskProxyless"] = "TurnstileTaskProxyless";
    TaskType["ComplexImageTask"] = "ComplexImageTask";
    TaskType["CustomTask"] = "CustomTask";
    TaskType["AmazonTask"] = "AmazonTask";
    TaskType["AmazonTaskProxyless"] = "AmazonTaskProxyless";
    TaskType["BinanceTaskProxyless"] = "BinanceTaskProxyless";
    TaskType["BinanceTask"] = "BinanceTask";
})(TaskType = exports.TaskType || (exports.TaskType = {}));
