// @zennolab_com/capmonstercloud-client/dist/Requests/AmazonRequestBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonRequestBase = void 0;
const CaptchaRequestBase_1 = require("./CaptchaRequestBase");
/**
 * Base GeeTest recognition request
 */
class AmazonRequestBase extends CaptchaRequestBase_1.CaptchaRequestBase {
    constructor({ type, nocache, websiteURL, challengeScript, captchaScript, websiteKey, context, iv, cookieSolution }) {
        super({ type, nocache });
        /**
         * By default false. If you need to use cookies "aws-waf-token", specify the value true. Otherwise, what you will get in return is "captcha_voucher" and "existing_token".
         */
        this.cookieSolution = false;
        this.websiteURL = websiteURL;
        this.challengeScript = challengeScript;
        this.captchaScript = captchaScript;
        this.websiteKey = websiteKey;
        this.context = context;
        this.iv = iv;
        this.cookieSolution = cookieSolution;
    }
}
exports.AmazonRequestBase = AmazonRequestBase;
